rpmalloc/rpmalloc.c

```diff
@@ -2380,8 +2380,17 @@ _rpmalloc_deallocate_direct_small_or_medium(span_t* span, void* block) {
 	--span->used_count;
 	span->free_list = block;
 	if (UNEXPECTED(span->used_count == span->list_size)) {
-		_rpmalloc_span_double_link_list_remove(&heap->size_class[span->size_class].partial_span, span);
-		_rpmalloc_span_release_to_cache(heap, span);
+		// Make sure we have synchronized the deferred list and list size by using acquire semantics
+		void* free_list;
+		do {
+			free_list = atomic_exchange_ptr_acquire(&span->free_list_deferred, INVALID_POINTER);
+		} while (free_list == INVALID_POINTER);
+		int all_free = (span->used_count == span->list_size);
+		atomic_store_ptr_release(&span->free_list_deferred, free_list);
+		if (all_free) {
+			_rpmalloc_span_double_link_list_remove(&heap->size_class[span->size_class].partial_span, span);
+			_rpmalloc_span_release_to_cache(heap, span);
+		}
```

Comment:
I've given this more thought and I think there is a need for an even stronger check, not less.

The free_list_deferred acts as the synchronization primitve to guarantee that only one thread can access the span after restoring the free_list_deferred pointer - the check for "all free" must happen while holding free_list_deferred in _rpmalloc_deallocate_defer_small_or_medium as well. And _rpmalloc_deallocate_direct_small_or_medium must always synchronize on free_list_deferred to make sure any external thread has completed the deferred free and will not access span again.

Otherwise the external thread could be preempted during the holding of the free_list_deferred pointer, the heap owner thread gets scheduled and releases the span, another thread gets scheduled and picks up the span, causing the span block_count to change, and finally the first external thread gets scheduled again and compares against the wrong span->block_count --> disaster.

I'll update the PR with the required change.

Reply:
I've given this more thought and I think there is a need for an even stronger check, not less.

The free_list_deferred acts as the synchronization primitve to guarantee that only one thread can access the span after restoring the free_list_deferred pointer - the check for "all free" must happen while holding free_list_deferred in _rpmalloc_deallocate_defer_small_or_medium as well. And _rpmalloc_deallocate_direct_small_or_medium must always synchronize on free_list_deferred to make sure any external thread has completed the deferred free and will not access span again.

Otherwise the external thread could be preempted during the holding of the free_list_deferred pointer, the heap owner thread gets scheduled and releases the span, another thread gets scheduled and picks up the span, causing the span block_count to change, and finally the first external thread gets scheduled again and compares against the wrong span->block_count --> disaster.

I'll update the PR with the required change.

Reply:
The question is then if it is safe to do the `if (span->used_count)` optimization you suggested - i.e only do the synchronization if there are deferred free blocks. It seems to me it should be, since if there are no used blocks (and `used_count` is only ever accessed by the owner thread) there can be no external thread executing in the _rpmalloc_deallocate_defer_small_or_medium function.